WIP rewrite http client docs (#961)
SAP/cloud-sdk: docs-js/features/connectivity/generic-http-client.mdx

---

> title: Perform HTTP Requests

Reviewer: [q] You also created a samples repo - are they mentioned here? If they are merged they sould be linked here.

Author: yes in line 49

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> This page explains how you can perform arbitrary HTTP requests using the SAP Cloud SDK.

Reviewer: ```suggestion
This guide explains how you can perform arbitrary HTTP requests using the SAP Cloud SDK.
```

> The HTTP client provided by the SAP Cloud SDK provides some convenience features which are explained in detail on this page.

Reviewer: ```suggestion
The HTTP client provided by the SAP Cloud SDK provides some convenience features which are explained in detail in this guide.
```

<!-- Prettier doesn't change this, cf https://docusaurus.io/docs/markdown-features/admonitions#usage-with-prettier -->
:::tip

> If you want to use an OData API, the [Executing a request using a generated OData client](../odata/execute-odata-request.mdx) section is the right place to look.

Reviewer: this sounds a bit informal

:::

## When to Use What?

> If the [OData client](../odata/execute-odata-request.mdx) can satisfy your requirements, it is always recommended to use that client directly as it provides the best convenience for interacting with OData APIs.

Reviewer: [q] What about OpenAPi clients?

> In case the OData client does not satisfy your requirements, the next thing you should look into is the HTTP client which is explained in this page.

Reviewer: ```suggestion
In case the OData client does not satisfy your requirements, the next thing you should look into is the HTTP client which is explained in this guide.
```

It is based on the open-source [axios HTTP client](https://axios-http.com) but provides convenience features which are specific to working on SAP Business Technology Platform.

> If the provided HTTP client does also not work for your use-case, we recommend using the [axios client](https://axios-http.com/docs/example) directly.

Reviewer: [q] Perhaps also mention opening an issue, because we want to support everything in general.

## `executeHttpRequest`

The most basic way to use the SAP Cloud SDK HTTP client is the function <LinkToLatestJsApiDocs slug="functions/sap_cloud_sdk_http_client.executeHttpRequest" name="executeHttpRequest" />.

```ts
executeHttpRequest(destination, requestConfig, httpRequestOptions): Promise<HttpResponse>;
```

<!-- Prettier doesn't change this, cf https://docusaurus.io/docs/markdown-features/admonitions#usage-with-prettier -->
:::tip

> This page provides code snippets, fully executable example code can be found in the [SAP-samples/cloud-sdk-js repository on GitHub](https://github.com/SAP-samples/cloud-sdk-js/tree/main/samples/http-client-examples).

Reviewer: ```suggestion
This guide provides code snippets, fully executable example code can be found in the [SAP-samples/cloud-sdk-js repository on GitHub](https://github.com/SAP-samples/cloud-sdk-js/tree/main/samples/http-client-examples).
```

:::

In the most simple form, this can be invoked like this:

```ts
const response = await executeHttpRequest(

>     {

Author: This is how prettier formatted the code with the default config.. not sure what you mean, is it that the object starts on a new line?

        url: "https://httpbin.org/get"
    },
    {
        method: 'get'
    },

>     {

Reviewer 2: [req] The last { } is not needed as its optional anyway.

- `httpRequestOptions` is of type `HttpRequestOptions`
  - This parameter is optional and can be left out
  - Can be used to enable [automatic CSRF token fetching](#csrf-token-fetching)

<!-- Prettier doesn't change this, cf https://docusaurus.io/docs/markdown-features/admonitions#usage-with-prettier -->
:::caution

Note that the `destination` and the `requestConfig` parameter have an overlap in the keys they accept.
For example, both have an `url` and an `authorization` property.
For keys that exist for both objects, the `requestConfig` takes precedence.

:::

The `executeHttpRequest` function returns a promise of <LinkToLatestJsApiDocs slug="interfaces/sap_cloud_sdk_http_client.HttpResponse" name="HttpResponse" />.

### CSRF Token Fetching

[Cross-Site Request Forgery (CSRF)](https://en.wikipedia.org/wiki/Cross-site_request_forgery) is a common type of attack on web applications.
The HTTP client provided by SAP Cloud SDK allows to automatically retrieve a CSRF token for HTTP APIs which support this.

> Fetching CSRF token is automatically enabled for all types of requests except HTTP `GET` requests.

Reviewer: ```suggestion
Fetching CSRF token is automatically enabled for all requests changing data e.g. `POST`, `DELETE` or `PATCH` requests.
```

Fetching CSRF token is not available for HTTP `GET` requests.

> Can be enabled or disabled via the `options` parameter of `executeHttpRequest`:

Reviewer: [q] what can be endabled?

```ts

>   {

Reviewer 2: [req] The indentation is off here as well.

    fetchCsrfToken: true,
  }
```

### Parameter Encoding

[URL encoding](https://en.wikipedia.org/wiki/Percent-encoding) allows using characters as part of the URL which normally would cause issues such as spaces.
For example, the space character is replaced with `%20`.

The HTTP client automatically takes care of encoding parameters except for *custom parameters*.

> For example, given this `requestConfig` object, the `requestConfig` parameters will be automatically escaped, but the `custom` will not.

Reviewer: [pp] Perhaps give a reason behind this choice: We want to give you a way to enforce certain values because sometimes backends behave strange.

```ts
{
  method: "get",
  params: {
    custom: {
      customParam: "a/b c",
    },
    requestConfig: {
      requestParam: "a/b c",
    },
  }
}
```

#### Customized Parameter Encoding

If you need to customize parameter encoding, you can provide your own function to take care of it like in this example.

> It uses the [encodeURI](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI) function, but you can do the transformation according to your own needs.

Reviewer: [pp] I found the `encodeURI` misleading. I would just put a `//put you custom encoding logic here` in the code sample.

Author: My goal was to provide a functioning example here, but as we have the example over in the samples repo maybe we don't need that

```ts
const myCustomParameterEncodingFunction: ParameterEncoder = function (
  params: Record<string, any>
): Record<string, any> {
  const encodedParams: Record<string, any> = {};

  for (const k in params) {
    // Customize your required encoding logic here
    encodedParams[k] = params[k].toString();
  }

  return encodedParams;
};

const encodingResponse = await executeHttpRequest(
  {
    url: "http://localhost:3000/encoding",
  },
  {
    method: "get",
    params: {
      custom: {
        customParam: "a/b c",
      },
      requestConfig: {
        requestParam: "a/b c",
      },
    },
    // Pass your custom encoding fuction
    parameterEncoder: myCustomParameterEncodingFunction
  },

>   {}

Reviewer 2: [req] {} not required

);
```

## `executeHttpRequestWithOrigin`

`executeHttpRequestWithOrigin` is a variation of `executeHttpRequest` which allows more fine-grained control over configuration precedence.

<!-- Prettier doesn't change this, cf https://docusaurus.io/docs/markdown-features/admonitions#usage-with-prettier -->
:::caution

> The term "origin" as it is used in the context of this page does **not** refer to [an URL as you might expect](https://developer.mozilla.org/en-US/docs/Glossary/Origin).

Reviewer: [pp] the as you might expect feels a bit strong.

Instead it refers to the *origin* or *source* of HTTP client configuration.

:::

> This function may be helpful in situations where you want to define the priority of headers and parameters, for example if``:

Reviewer 2: ```suggestion
This function may be helpful in situations where you want to define the priority of headers and parameters, for example if:
```

> - You want to enforce a hard-coded API key/JWT and ignore the ones from e.g. the destination

Reviewer 2: ```suggestion
- You want to enforce a hard-coded API key/JWT and ignore the ones from, e.g., the destination
```

- You want a default query parameter, which can be overwritten by the destination properties

Headers and query parameters with the same name are prioritized in the following order:
1. `custom`
2. Destination related headers/query parameters
3. `requestConfig`.

> In this example the `custom` `apiKey` and `myParam` override the request config:

Reviewer 2: ```suggestion
In this example, the `custom` `apiKey` and `myParam` override the request config:
```

  }
}
```

#### Customized Parameter Encoding

If you need to customize parameter encoding, you can provide your own function to take care of it like in this example.

```ts
const myCustomParameterEncodingFunction: ParameterEncoder = function (
  params: Record<string, any>
): Record<string, any> {
  const encodedParams: Record<string, any> = {};

  for (const k in params) {
    // Customize your required encoding logic here
    encodedParams[k] = params[k].toString();
  }

  return encodedParams;
};

const encodingResponse = await executeHttpRequest(
  {
    url: "http://localhost:3000/encoding",
  },
  {
    method: "get",
    params: {
      custom: {
        customParam: "a/b c",
      },
      requestConfig: {
        requestParam: "a/b c",
      },
    },
    // Pass your custom encoding fuction
    parameterEncoder: myCustomParameterEncodingFunction
  },
  {}
);
```

## `executeHttpRequestWithOrigin`

`executeHttpRequestWithOrigin` is a variation of `executeHttpRequest` which allows more fine-grained control over configuration precedence.

<!-- Prettier doesn't change this, cf https://docusaurus.io/docs/markdown-features/admonitions#usage-with-prettier -->
:::caution

> The term "origin" as it is used in the context of this guide does **not** refer to [an URL as in the "same-origin policy"](https://developer.mozilla.org/en-US/docs/Glossary/Origin).

Reviewer 2: ```suggestion
The term "origin" as it is used in the context of this guide does **not** refer to [a URL as in the "same-origin policy"](https://developer.mozilla.org/en-US/docs/Glossary/Origin).
```
